apiserver/facades/agent/uniter/networkinfoiaas.go

```diff
@@ -188,82 +207,70 @@ func (n *NetworkInfoIAAS) populateMachineNetworkInfos() error {
 		}
 	}
 
-	// Link-layer devices are set in a single transaction for all devices
-	// observed on the machine, so the first result will include them all.
-	var addresses []*state.Address
-	retryArg := n.retryFactory()
-	retryArg.Func = func() error {
-		var err error
-		addresses, err = machine.AllAddresses()
-		return err
+	// This is not ideal. We need device information associated with the state
+	// Address representation, but we also need the addresses in a form that
+	// can be sorted for scope and primary/secondary status.
+	// We create a map for the information we need to return, and a separate
+	// sorted slice for iteration in the correct order.
+	addrs, err := machine.AllAddresses()
+	if err != nil {
+		n.populateMachineNetworkInfoErrors(spaceSet, err)
+		return nil
 	}
-	retryArg.IsFatalError = func(err error) bool {
-		return err != nil
+	addrByIP := make(map[string]*state.Address)
+	for _, addr := range addrs {
+		addrByIP[addr.Value()] = addr
 	}
-	if err := retry.Call(retryArg); err != nil {
-		result := params.NetworkInfoResult{Error: apiservererrors.ServerError(
-			errors.Annotate(err, "getting devices addresses"))}
-		for _, id := range spaceSet.Values() {
-			if _, ok := results[id]; !ok {
-				results[id] = result
-			}
-		}
-		n.machineNetworkInfos = results
+
+	spaceAddrs, err := machine.AllDeviceSpaceAddresses()
+	if err != nil {
+		n.populateMachineNetworkInfoErrors(spaceSet, err)
 		return nil
 	}
+	network.SortAddresses(spaceAddrs)
 
-	logger.Debugf("Looking for address from %v in spaces %v", addresses, spaceSet.Values())
+	logger.Debugf("Looking for address from %v in spaces %v", spaceAddrs, spaceSet.Values())
 
 	var privateLinkLayerAddress *state.Address
-	for _, addr := range addresses {
-		subnet, err := addr.Subnet()
-		switch {
-		case errors.IsNotFound(err):
-			logger.Debugf("skipping %s: not linked to a known subnet (%v)", addr, err)
+	for _, spaceAddr := range spaceAddrs {
+		addr, ok := addrByIP[spaceAddr.Value]
+		if !ok {
+			return errors.Errorf("address representations inconsistent; could not find %s", spaceAddr.Value)
+		}
+
+		if spaceAddr.SpaceID == "" {
+			logger.Debugf("skipping %s: not linked to a known space.", spaceAddr)
 
 			// For a space-less model, we will not have subnets populated,
-			// and will therefore not find a subnet for the address.
+			// and will therefore not find a space for the address.
 			// Capture the link-layer information for machine private address
 			// so that we can return as much information as possible.
 			// TODO (manadart 2020-02-21): This will not be required once
 			// discovery (or population of subnets by other means) is
-			// introduced for the non-space IAAS providers (LXD, manual, etc).
-			if addr.Value() == privateIPAddress {
+			// introduced for the non-space IAAS providers (vSphere etc).
+			if spaceAddr.Value == privateIPAddress {
```

Comment:
Is vsphere the only remaining provider without subnet discovery?

Reply:
Pretty sure it is.